safehttp/plugins/csp/csp.go

```diff
@@ -0,0 +1,170 @@
+//
+// 	https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package csp
+
+import (
+	"crypto/rand"
+	"encoding/base64"
+	"fmt"
+	"strings"
+
+	"github.com/google/go-safeweb/safehttp"
+)
+
+// Directive is the name of a single CSP directive.
+type Directive string
+
+const (
+	DirectiveScriptSrc Directive = "script-src"
+	DirectiveStyleSrc            = "style-src"
+	DirectiveObjectSrc           = "object-src"
+	DirectiveBaseURI             = "base-uri"
+	DirectiveReportURI           = "report-uri"
+)
+
+const (
+	ValueHTTPS         = "https:"
+	ValueHTTP          = "http:"
+	ValueUnsafeEval    = "'unsafe-eval'"
+	ValueUnsafeInline  = "'unsafe-inline'"
+	ValueNone          = "'none'"
+	ValueStrictDynamic = "'strict-dynamic'"
+)
+
+// PolicyDirective contains a single CSP directive.
+type PolicyDirective struct {
+	Directive Directive
+	Values    []string
+	AddNonce  bool
+}
+
+// nonceSize is the size of the nonces in bytes.
+const nonceSize = 8
+
+func generateNonce(readRand func([]byte) (int, error)) string {
+	if readRand == nil {
+		readRand = rand.Read
+	}
+	b := make([]byte, nonceSize)
+	_, err := readRand(b)
+	if err != nil {
+		// TODO: handle this better, what should happen here?
+		panic(err)
+	}
+	return base64.StdEncoding.EncodeToString(b)
+}
+
+// Policy defines a CSP policy, containing many directives.
+type Policy struct {
+	Directives []*PolicyDirective
+	// readRand is used for dependency injection in tests.
+	readRand func([]byte) (int, error)
+}
+
+// NewPolicy creates a new strict, nonce-based CSP.
+// See https://csp.withgoogle.com/docs/strict-csp.html for more info.
+//
+// TODO: maybe reportURI should be safehttp.URL?
+func NewPolicy(reportURI string) *Policy {
+	return &Policy{
+		Directives: []*PolicyDirective{
+			{Directive: DirectiveObjectSrc, Values: []string{ValueNone}, AddNonce: false},
+			{
+				Directive: DirectiveScriptSrc,
+				Values: []string{
+					ValueUnsafeInline,
+					ValueUnsafeEval,
+					ValueStrictDynamic,
+					ValueHTTPS,
+					ValueHTTP,
+				},
+				AddNonce: true,
+			},
+			{Directive: DirectiveBaseURI, Values: []string{ValueNone}, AddNonce: false},
+			{Directive: DirectiveReportURI, Values: []string{reportURI}, AddNonce: false},
+		},
+	}
+}
+
+// Serialize serializes this policy for use in a Content-Security-Policy header
+// or in a Content-Security-Policy-Report-Only header. The nonces generated for
+// each directive are also returned.
+func (p Policy) Serialize() (csp string, nonces map[Directive]string) {
+	nonces = make(map[Directive]string)
+	values := make([]string, 0, len(p.Directives))
+
+	for _, d := range p.Directives {
+		var b strings.Builder
```

Comment:
Use one builder and reset it in the last line of the for loop. Or even better since you strings.Join at the end just append what is needed on each iteration except the last.

Reply:
I wanted to avoid conditionals for iteration numbers, this is why I suggested `strings.Join`. Unless I misunderstood your suggestion.